share/man/man8/powertop.8
'\" te
.\" Copyright (c) 2009, Sun Microsystems, Inc. All Rights Reserved
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License"). You may not use this file except in compliance with the License. You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.
.\"  See the License for the specific language governing permissions and limitations under the License. When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE. If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH POWERTOP 8 "May 13, 2017"
.SH NAME
powertop \- report and analyze events that affect power management
.SH SYNOPSIS
.LP
.nf
\fBpowertop\fR [\fB-c\fR \fIprocessor_id\fR] [\fB-d\fR \fIcount\fR] [\fB-t\fR \fIinterval\fR] [\fB-v\fR] [\fB-h\fR]
.fi

.SH DESCRIPTION
.LP
\fBPowerTOP\fR is an observability tool that shows how effectively the system
is taking advantage of the CPU's power management features. By running the tool
on an otherwise idle system, the user can see for how long the CPU is running
at different power states. Ideally, an unutilized (idle) system spends 100%
of its time running at the lowest power state, but because of background user
and kernel activity (random software periodically waking to poll status), idle
systems can consume more power than they should.
.sp
.LP
The tool analyzes system activity periodically and displays a summary of how
long the processor is executing at each supported power state. It also displays
the top activities responsible for causing the CPU to wake up and use more
energy. This report allows the user to identify and diagnose problematic areas
of the system and optimize its power efficiency.
.sp
.LP
\fBPowerTOP\fR averages the amount of activity that is preventing the CPU from
entering a lower power state and presents it on the "Wakeups-from-idle per
second" field. This value represents the total number of wake-ups divided by
the current interval. Notice that not all events are displayed on the screen at
all times.
.sp
.LP
During execution, a user can force a refresh of the analysis by pressing the
\fBR\fR key. The interval time is restored to the default or to a specified
value. To quit the application, the user must press the \fBQ\fR key.
.sp
.LP
If you are running as root (superuser) or in the Primary Administrator role,
the tool makes suggestions as how the system can be improved from a power
management perspective.
.sp
.LP
PowerTOP allows you to freeze each of its three subwindows, enabling you to
better analyze each subwindow's contents.  When you have three different DTrace
scripts running at the same time, freezing subwindows lowers CPU consumption.
This feature is activated by pressing the \fBi\fR, \fBf\fR, or \fBe\fR key
while the tool running.  Pressing one of these keys freezes the idle,
frequency, or event report, respectively. Pressing it once more, unfreezes it.
This feature is most useful when invoked while the application being analyzed
is running.
.sp
.LP
\fBPowerTOP\fR runs on \fBxVM\fR(5) domains. However, the report for idle state
transitions might or might not be accurate as the physical CPU can be shared by
different virtual CPUs. Both wakeup count and event report displays information
regarding the current virtualized environment.
.SH OPTIONS
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-c\fR [\fIprocessor_id\fR]\fR
.ad
.sp .6
.RS 4n
Specifies which CPU the tool should observe.
.RE

.sp
.ne 2
.na
\fB\fB-d\fR [\fIcount\fR]\fR
.ad
.sp .6
.RS 4n
Dumps the results of \fIcount\fR analysis of system activity to the screen.
.RE

.sp
.ne 2
.na
\fB\fB-h\fR\fR
.ad
.sp .6
.RS 4n
Displays the command's usage.
.RE

.sp
.ne 2
.na
\fB\fB-t\fR [\fIinterval\fR]\fR
.ad
.sp .6
.RS 4n
Specifies the interval, in seconds, at which the tool analyzes the system. The
possible values are between 1 and 100; the default is 5 seconds.
.RE

.sp
.ne 2
.na
\fB\fB-v\fR\fR
.ad
.sp .6
.RS 4n
Switches to verbose mode, including noting firings of the kernel cyclic
subsystem in the event report.
.RE

.SH EXAMPLES
.LP
\fBExample 1 \fRSetting the Interval
.sp
.LP
The following command sets the interval to two seconds.

.sp
.in +2
.nf
% \fBpowertop -t 2\fR
.fi
.in -2
.sp

.LP
\fBExample 2 \fRAnalyzing and Dumping System Activity
.sp
.LP
The following command analyzes and dumps system activity to the standard output
four times.

.sp
.in +2
.nf
% \fBpowertop -d 4\fR
.fi
.in -2
.sp

.LP
\fBExample 3 \fRReporting Cyclic Subsystem Activity
.sp
.LP
The following command reports cyclic subsystem activity.

.sp
.in +2
.nf
% \fBpowertop -v\fR
.fi
.in -2
.sp

.LP
\fBExample 4 \fRAnalyzing Activity on a Specific Processor
.sp
.LP
The following command runs PowerTOP and only displays data for CPU 3:

.sp
.in +2
.nf
% \fBpowertop -c 3\fR
.fi
.in -2
.sp

.SH EXIT STATUS
.ne 2
.na
\fB\fB0\fR\fR
.ad
.sp .6
.RS 4n
Successful operation.
.RE

.sp
.ne 2
.na
\fB\fB1\fR\fR
.ad
.sp .6
.RS 4n
An error occurred.
.RE

.sp
.ne 2
.na
\fB\fB2\fR\fR
.ad
.sp .6
.RS 4n
Incorrect usage.
.RE

.SH ATTRIBUTES
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
Architecture	x86, SPARC
_
Interface Stability	Volatile
.TE

.SH SEE ALSO
.LP
\fBkstat\fR(8), \fBpmconfig\fR(8), \fBpowerd\fR(8), \fBpsrinfo\fR(8),
\fBuadmin\fR(2), \fBlibdevinfo\fR(3LIB), \fBattributes\fR(5), \fBxVM\fR(5),
\fBcpr\fR(7), \fBpm\fR(7D), \fBpm-components\fR(9P), \fBremovable-media\fR(9P)
.SH USAGE
.LP
You must have \fBDTrace\fR privileges to run \fBPowerTOP\fR and root
(superuser) privileges or assume the Primary Administrator role for the tool to
suggest improvements to the system.


usr/src/cmd/powertop/common/powertop.c
/*
 * Copyright 2009, Intel Corporation
 * Copyright 2009, Sun Microsystems, Inc
 *
 * This file is part of PowerTOP
 *
 * This program file is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program in a file named COPYING; if not, write to the
 * Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA
 *
 * Authors:
 *	Arjan van de Ven <arjan@linux.intel.com>
 *	Eric C Saxe <eric.saxe@sun.com>
 *	Aubrey Li <aubrey.li@intel.com>
 */

/*
 * GPL Disclaimer
 *
 * For the avoidance of doubt, except that if any license choice other
 * than GPL or LGPL is available it will apply instead, Sun elects to
 * use only the General Public License version 2 (GPLv2) at this time
 * for any software where a choice of GPL license versions is made
 * available with the language indicating that GPLv2 or any later
 * version may be used, or where a choice of which version of the GPL
 * is applied is otherwise unspecified.
 */

#include <getopt.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <ctype.h>
#include <poll.h>
#include "powertop.h"

/*
 * Global variables, see powertop.h for comments and extern declarations.
 * These are ordered by type, grouped by usage.
 */
int			g_bit_depth;
int 			g_total_events, g_top_events;
int			g_npstates, g_max_cstate, g_longest_cstate;
uint_t			g_features;
uint_t			g_ncpus;
uint_t			g_ncpus_observed;

processorid_t 		*g_cpu_table;

double			g_interval_length;
hrtime_t		g_total_c_time;

uchar_t			g_op_mode;
boolean_t		g_gui;
uint_t			g_observed_cpu;

event_info_t    	g_event_info[EVENT_NUM_MAX];
state_info_t		g_cstate_info[NSTATES];
freq_state_info_t	g_pstate_info[NSTATES];
cpu_power_info_t	*g_cpu_power_states;

boolean_t		g_turbo_supported;
boolean_t		g_sig_resize;

uint_t			g_argc;
char			**g_argv;

static const int	true = 1;

void
pt_sig_handler(int sig)
{
	switch (sig) {
	case SIGWINCH:
		g_sig_resize = B_TRUE;
		break;
	}
}

int
main(int argc, char **argv)
{
	double		interval, interval_usr;
	hrtime_t 	interval_start;
	int		index2 = 0, c, dump_count = 0;
	char		*endptr, key;
	boolean_t	root_user = B_FALSE;
	struct pollfd	pollset;

	static struct option opts[] = {
		{ "dump", 1, NULL, 'd' },
		{ "time", 1, NULL, 't' },
		{ "help", 0, NULL, 'h' },
		{ "cpu", 1, NULL, 'c' },
		{ "verbose", 0, NULL, 'v' },
		{ 0, 0, NULL, 0 }
	};

	pt_set_progname(argv[0]);

	/*
	 * Enumerate the system's CPUs, populate cpu_table, g_ncpus
	 */
	if ((g_ncpus = g_ncpus_observed = pt_enumerate_cpus()) == 0)
		exit(EXIT_FAILURE);

	if ((g_bit_depth = pt_get_bit_depth()) < 0)
		exit(EXIT_FAILURE);

	g_features = 0;
	interval = interval_usr = INTERVAL_DEFAULT;
	g_op_mode = PT_MODE_DEFAULT;
	g_max_cstate = 0;
	g_argv = NULL;
	g_argc = 0;
	g_observed_cpu = 0;
	g_turbo_supported = B_FALSE;
	g_sig_resize = B_FALSE;
	g_curr_sugg = NULL;

	while ((c = getopt_long(argc, argv, "d:t:hvc:", opts, &index2))
	    != EOF) {
		if (c == -1)
			break;

		switch (c) {
		case 'd':
			if (PT_ON_DUMP) {
				pt_usage();
				exit(EXIT_USAGE);
			}

			g_op_mode |= PT_MODE_DUMP;
			g_gui = B_FALSE;
			dump_count = (int)strtod(optarg, &endptr);

			if (dump_count <= 0 || *endptr != '\0') {
				pt_usage();
				exit(EXIT_USAGE);
			}

			break;
		case 't':
			if (PT_ON_TIME) {
				pt_usage();
				exit(EXIT_USAGE);
			}

			g_op_mode |= PT_MODE_TIME;
			interval = interval_usr = (double)strtod(optarg,
			    &endptr);

			if (*endptr != '\0' || interval < 1 ||
			    interval > INTERVAL_MAX) {
				pt_usage();
				exit(EXIT_USAGE);
			}

			break;
		case 'v':
			if (PT_ON_CPU || PT_ON_VERBOSE) {
				pt_usage();
				exit(EXIT_USAGE);
			}

			g_op_mode |= PT_MODE_VERBOSE;
			break;
		case 'c':
			if (PT_ON_CPU || PT_ON_VERBOSE) {
				pt_usage();
				exit(EXIT_USAGE);
			}

			g_op_mode |= PT_MODE_CPU;
			g_observed_cpu = (uint_t)strtod(optarg, &endptr);

			if (g_observed_cpu >= g_ncpus) {
				pt_usage();
				exit(EXIT_USAGE);
			}

			g_argc = 1;
			g_ncpus_observed = 1;

			if ((g_argv = malloc(sizeof (char *))) == NULL)
				return (EXIT_FAILURE);

			if ((*g_argv = malloc(sizeof (char) * 5)) == NULL)
				return (EXIT_FAILURE);

			(void) snprintf(*g_argv, 5, "%d\0", g_observed_cpu);
			break;
		case 'h':
			pt_usage();
			exit(EXIT_SUCCESS);
		default:
			pt_usage();
			exit(EXIT_USAGE);
		}
	}

	if (optind < argc) {
		pt_usage();
		exit(EXIT_USAGE);
	}

	(void) printf("%s   %s\n\n", TITLE, COPYRIGHT_INTEL);

	(void) printf("Collecting data for %.2f second(s) \n",
	    (float)interval);

	/* Prepare P-state statistics */
	if (pt_cpufreq_stat_prepare() == 0)
		g_features |= FEATURE_PSTATE;

	/* Prepare C-state statistics */
	if (pt_cpuidle_stat_prepare() == 0)
		g_features |= FEATURE_CSTATE;
	else
		/*
		 * PowerTop was unable to run a DTrace program,
		 * most likely for lack of permissions.
		 */
		exit(EXIT_FAILURE);

	/* Prepare event statistics */
	if (pt_events_stat_prepare() != -1)
		g_features |= FEATURE_EVENTS;

	/*
	 * If the system is running on battery, find out what's
	 * the kstat module for it
	 */
	pt_battery_mod_lookup();

	/* Prepare turbo statistics */
	if (pt_turbo_stat_prepare() == 0)
		g_features |= FEATURE_TURBO;

	/*
	 * Initialize the display.
	 */
	if (!PT_ON_DUMP) {
		pt_display_init_curses();
		pt_display_setup(B_FALSE);
		(void) signal(SIGWINCH, pt_sig_handler);

		pt_display_title_bar();
		pt_display_status_bar();

		g_gui = B_TRUE;
		pollset.fd = STDIN_FILENO;
		pollset.events = POLLIN;
	}

	/*
	 * Installs the initial suggestions, running as root and turning CPU
	 * power management ON.
	 */
	if (geteuid() != 0) {
		pt_sugg_as_root();
	} else {
		root_user = B_TRUE;
		pt_cpufreq_suggest();
	}

	while (true) {
		key = 0;

		if (g_sig_resize)
			pt_display_resize();

		interval_start = gethrtime();

		if (!PT_ON_DUMP) {
			if (poll(&pollset, (nfds_t)1,
			    (int)(interval * MILLISEC)) > 0)
				(void) read(STDIN_FILENO, &key, 1);
		} else {
			(void) sleep((int)interval);
		}

		g_interval_length = (double)(gethrtime() - interval_start)
		    /NANOSEC;

		g_top_events = 0;
		g_total_events = 0;

		(void) memset(g_event_info, 0,
		    EVENT_NUM_MAX * sizeof (event_info_t));
		(void) memset(g_cstate_info, 0,
		    NSTATES * sizeof (state_info_t));

		/* Collect idle state transition stats */
		if (g_features & FEATURE_CSTATE &&
		    pt_cpuidle_stat_collect(g_interval_length) < 0) {
			/* Reinitialize C-state statistics */
			if (pt_cpuidle_stat_prepare() != 0)
				exit(EXIT_FAILURE);

			continue;
		}

		/* Collect frequency change stats */
		if (g_features & FEATURE_PSTATE &&
		    pt_cpufreq_stat_collect(g_interval_length) < 0) {
			/* Reinitialize P-state statistics */
			if (pt_cpufreq_stat_prepare() != 0)
				exit(EXIT_FAILURE);

			continue;
		}

		/* Collect event statistics */
		if (g_features & FEATURE_EVENTS &&
		    pt_events_stat_collect() < 0) {
			/* Reinitialize event statistics */
			if (pt_events_stat_prepare() != 0)
				exit(EXIT_FAILURE);

			continue;
		}

		/* Collect turbo statistics */
		if (g_features & FEATURE_TURBO &&
		    pt_turbo_stat_collect() < 0)
			exit(EXIT_FAILURE);

		/* Show CPU power states */
		pt_display_states();

		/* Show wakeups events affecting PM */
		if (g_features & FEATURE_EVENTS) {
			pt_display_wakeups(g_interval_length);
			pt_display_events(g_interval_length);
		}

		pt_battery_print();

		if (key && !PT_ON_DUMP) {
			switch (toupper(key)) {
			case 'Q':
				exit(EXIT_SUCCESS);
				break;

			case 'R':
				interval = 3;
				break;
			}

			/*
			 * Check if the user has activated the current
			 * suggestion.
			 */
			if (g_curr_sugg != NULL &&
			    toupper(key) == g_curr_sugg->key &&
			    g_curr_sugg->func)
				g_curr_sugg->func();
		}

		if (dump_count)
			dump_count--;

		/* Exits if user requested a dump */
		if (PT_ON_DUMP && !dump_count)
			exit(EXIT_SUCCESS);

		/* No key pressed, will suggest something */
		if (!key && !dump_count)
			pt_sugg_pick();

		/* Refresh display */
		if (!PT_ON_DUMP)
			pt_display_update();

		if (root_user)
			pt_cpufreq_suggest();

		/*
		 * Update the interval based on how long the CPU was in the
		 * longest c-state during the last snapshot. If the user
		 * specified an interval we skip this bit and keep it fixed.
		 */
		if (g_features & FEATURE_CSTATE && !PT_ON_TIME &&
		    g_longest_cstate > 0 &&
		    g_cstate_info[g_longest_cstate].events > 0) {
			double deep_idle_res = (((double)
			    g_cstate_info[g_longest_cstate].total_time/MICROSEC
			    /g_ncpus)/g_cstate_info[g_longest_cstate].events);

			if (deep_idle_res < INTERVAL_DEFAULT ||
			    (g_total_events/interval) < 1)
				interval = INTERVAL_DEFAULT;
			else
				interval = INTERVAL_UPDATE(deep_idle_res);
		} else {
			/*
			 * Restore interval after a refresh.
			 */
			if (key)
				interval = interval_usr;
		}
	}

	return (EXIT_SUCCESS);
}
